File: src/etch/bindings/csharp/msg/ComboValidator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Etch.Msg
{
    /// <summary>
    /// Validates values using one of two validators
    /// </summary>
    public class ComboValidator : Validator
    {
        /// <summary>
        /// Constructs the combo validator
        /// </summary>
        /// <param name="a"></param>
        /// <param name="b"></param>
        public ComboValidator( Validator a, Validator b )
        {
            this.a = a;
            this.b = b;
        }

        private readonly Validator a;
        private readonly Validator b;

        public override string ToString()
        {
            return String.Format( "({0} OR {1})", a, b );
        }

        public override Validator ElementValidator()
        {
            Validator na;

            try
            {
                na = a.ElementValidator();
            }
            catch ( Exception )
            {
                na = null;
            }

            Validator nb;
            try
            {
                nb = b.ElementValidator();
            }
            catch ( Exception )
            {
                nb = null;
            }

            if ( na == null && nb == null )
                throw new ArgumentException( "na == null and nb == null" );

            if ( na == null )
                return nb;

            if ( nb == null )
                return na;

            return new ComboValidator( na, nb );

        }

        public override bool Validate( object value )
        {
            return a.Validate( value ) || b.Validate( value );
        }

      

        public override object ValidateValue(object value)
        {
            try
            {
                return a.ValidateValue(value);
            }
            catch (Exception)
            {
                return b.ValidateValue(value);
            }
        }

        public Validator A()
        {
            return a;
        }

        public Validator B()
        {
            return b;
        }
    }
}

File: src/etch/bindings/java/msg/ComboValidator.java
/* $Id$
 *
 * Created by wert on Aug 27, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.msg;

/**
 * Validates values using one of two validators.
 */
public class ComboValidator extends Validator
{
	/**
	 * Constructs the ComboValidator.
	 *
	 * @param a
	 * @param b
	 */
	public ComboValidator( Validator a, Validator b )
	{
		this.a = a;
		this.b = b;
	}
	
	private final Validator a;
	
	private final Validator b;
	
	@Override
	public String toString()
	{
		return String.format( "(%s) OR (%s)", a, b );
	}

	@Override
	public Validator elementValidator()
	{
		Validator na = a.elementValidator();
		Validator nb = b.elementValidator();
		
		if (na == null && nb == null)
			return null;
		
		if (na == null)
			return nb;
		
		if (nb == null)
			return na;
		
		return new ComboValidator( na, nb );
	}

	@Override
	public boolean validate( Object value )
	{
		return a.validate( value ) || b.validate( value );
	}
	
	@Override
	public Byte checkValue( Object value )
	{
		Byte type = a.checkValue( value );
		if (type == null)
			type = b.checkValue( value );
		return type;
	}
}
